polar-core/src/normalize.rs

```diff
@@ -0,0 +1,317 @@
+                    self.clone_with_value(Expression(Operation { operator: Or, args }))
+                }
+                Or => {
+                    assert_eq!(args.len(), 2);
+                    let args = vec![args[0].negated(), args[1].negated()];
+                    self.clone_with_value(Expression(Operation {
+                        operator: And,
+                        args,
+                    }))
+                }
+                _ => self.clone_with_value(Expression(op!(Not, self.clone()))),
+            },
+            _ => self.clone_with_value(Expression(op!(Not, self.clone()))),
+        }
+    }
+
+    /// binary tree normal form -- all and / or nodes form a binary tree
+    fn btnf(&self) -> Self {
+        use Operator::*;
+        match self.value().as_expression() {
+            Ok(Operation { operator, args }) if *operator == And || *operator == Or => {
+                match args.len() {
+                    // empty -> boolean
+                    0 => self.clone_with_value(Value::Boolean(*operator == And)),
+                    // one -> lift it out
+                    1 => args[0].btnf(),
+                    // else fold
+                    _ => args
+                        .iter()
+                        .map(|a| a.btnf())
+                        .reduce(|m, x| {
+                            self.clone_with_value(Value::Expression(Operation {
+                                operator: *operator,
+                                args: vec![m, x],
+                            }))
+                        })
+                        .unwrap(),
+                }
+            }
+            _ => self.clone(),
+        }
+    }
+
+    /// negation normal form -- all not operators are nested inside of and/or
+    /// using De Morgan's law
+    fn nnf(&self) -> Self {
+        use Operator::*;
+        match self.value().as_expression() {
+            Err(_) => self.clone(),
+            Ok(Operation {
+                operator: Not,
+                args,
+            }) => args[0].nnf().negated(),
+            Ok(Operation { operator, args }) => {
+                self.clone_with_value(Value::Expression(Operation {
+                    operator: *operator,
+                    args: args.iter().cloned().map(|t| t.nnf()).collect(),
+                }))
+            }
+        }
+    }
+
+    fn _hs(&self, n: usize) -> Self {
+        self.value().as_expression().unwrap().args[n].clone()
+    }
+
+    fn lhs(&self) -> Self {
+        self._hs(0)
+    }
+    fn rhs(&self) -> Self {
+        self._hs(1)
+    }
+
+    /// normalize an expression by fully distributing logical connectives. the
+    /// expression must already be in btnf and nnf. the arguments are two
+    /// predicate / constructor pairs:
+    /// - p1/c1: predicate/constructor for *inner* connective (and for dnf, or for cnf)
+    /// - p2/c2: predicate/constructor for *outer* connective (or for dnf, and for cnf)
+    fn distribute(
```

Comment:
Is there a reference you could link maybe about this operation? I haven't reviewed it for correctness or thought about edge cases at all -- assuming it's translated from somewhere / well known?

Reply:
No, it's just literally what you do when you distribute the connectives over each other 🤷‍♀️